concent_api/middleman/queue_operations.py

```diff
@@ -0,0 +1,186 @@
+    message_tracker: Dict[int, MessageTrackerItem],
+):
+    message_tracker = message_tracker  # type: OrderedDict[int, MessageTrackerItem]
+    while True:
+        try:
+            frame = await handle_frame_receive_async(signing_service_reader, settings.SIGNING_SERVICE_PUBLIC_KEY)
+        except IncompleteReadError:
+            logger.info(f"Signing Service has closed the connection. Ending 'response_producer' coroutine")
+            break
+        except (FrameInvalidMiddlemanProtocolError,
+                BrokenEscapingInFrameMiddlemanProtocolError,
+                SignatureInvalidMiddlemanProtocolError,
+                PayloadTypeInvalidMiddlemanProtocolError) as exception:
+            logger.info(f"Received invalid message: error code={map_exception_to_error_code(exception)}")
+            print(exception)
+            continue
+        logger.info(
+            f"Received message from Signing Service: request ID = {frame.request_id}"
+        )
+        if frame.request_id not in message_tracker:
+            logger.info(f"There is no entry in Message Tracker for request ID = {frame.request_id}, skipping...")
+            continue
+        current_track = message_tracker[frame.request_id]  # type: MessageTrackerItem
+        if current_track.connection_id not in response_queue_pool:
+            logger.info(f"Response queue for {current_track.connection_id} doesn't exist anymore, skipping...")
+            # there will be no more processing for current_track, it should be removed from message_tracker
+            del message_tracker[frame.request_id]
+            continue
+        discard_entries_for_lost_messages(frame.request_id, message_tracker, logger)
+        await response_queue_pool[current_track.connection_id].put(
+            ResponseQueueItem(
+                message=frame.payload,
+                concent_request_id=current_track.concent_request_id,
+                timestamp=now().timestamp(),
+            )
+        )
+        # message has been processed, now it should be removed from message_tracker
+        del message_tracker[frame.request_id]
+
+
+async def response_consumer(
+    response_queue: Queue,
+    writer: StreamWriter,
+    connection_id: int
+):
+    while True:
+        item = await response_queue.get()  # type: ResponseQueueItem
+        assert isinstance(item, ResponseQueueItem)
+        frame = create_middleman_protocol_message(
+            PayloadType.GOLEM_MESSAGE,
+            item.message,
+            item.concent_request_id,
+        )  # type: GolemMessageFrame
+        await send_over_stream_async(frame, writer, settings.CONCENT_PRIVATE_KEY)
+        logger.info(
+            f"Message (request ID={frame.request_id}) for Concent has been sent for connection ID={connection_id}"
+        )
+        response_queue.task_done()
+
+
+def discard_entries_for_lost_messages(
+    current_request_id: int,
+    message_tracker: Dict[int, MessageTrackerItem],
+    logger_: Logger
```

Comment:
Why not using the main logger ?

Reply:
quite frankly I could (the logger, that is actually passed to this function **is** the main logger), but having the opportunity to feed it directly made the function more flexible and easier for testing (dependency injection vs monkey patching). Anyway, if you insist, I will remove the parameter ;)

Reply:
No need, I was just curious why it is done this way, if it makes things simpler lets keep it.

Reply:
:+1: